src/librustc_middle/ty/query/plumbing.rs

```diff
@@ -234,18 +234,54 @@ macro_rules! hash_result {
 
 macro_rules! define_queries {
     (<$tcx:tt> $($category:tt {
-        $($(#[$attr:meta])* [$($modifiers:tt)*] fn $name:ident: $node:ident($K:ty) -> $V:ty,)*
+        $($(#[$attr:meta])* [$($modifiers:tt)*] fn $name:ident: $node:ident($($K:tt)*) -> $V:ty,)*
     },)*) => {
         define_queries_inner! { <$tcx>
-            $($( $(#[$attr])* category<$category> [$($modifiers)*] fn $name: $node($K) -> $V,)*)*
+            $($( $(#[$attr])* category<$category> [$($modifiers)*] fn $name: $node($($K)*) -> $V,)*)*
         }
     }
 }
 
+macro_rules! define_query_getter {
+    (TyCtxtAt<$tcx:tt>, $(#[$attr:meta])* $name:ident(DefId) -> $V:ty) => {
+        $(#[$attr])*
+        #[inline(always)]
+        pub fn $name(self, key: impl Into<DefId>) -> $V {
+            fn inner(this: TyCtxtAt<$tcx>, key: DefId) -> $V {
+                get_query::<queries::$name<'_>, _>(this.tcx, this.span, key)
+            }
```

Comment:
Is `get_query` inlined? This does nothing otherwise.

Reply:
This was an attempt to address [@Zoxc's comment](https://github.com/rust-lang/rust/pull/70956#discussion_r406293448). I don't understand why it's beneficial for the caller of `get_query` to be monomorphic, since it is `#[inline(never)]`. Perhaps I misunderstood?

Reply:
IFAIK, the LLVM codegen of the `{get,force,ensure}_query` functions is responsible for most of the compile time for librustc_middle. It happens once for each of the 170 queries. By making the caller generic, part of this codegen may be replicated in caller crates. This would have a serious impact on rustc's compile time. Please correct me if `#[inline(never)]` is enough to prevent this.

Reply:
The wrapper functions are already `#[inline(always)]` so they would be codegen'd cross-crate even when not generic.